contracts/shapella-upgrade/ShapellaUpgradeTemplate.sol

```diff
@@ -0,0 +1,506 @@
+
+        // TODO mainnet: hardcode the values
+        _locator = _config.locator;
+        _eip712StETH = _config.eip712StETH;
+        _voting = _config.voting;
+        _nodeOperatorsRegistry = _config.nodeOperatorsRegistry;
+        _hashConsensusForAccountingOracle = _config.hashConsensusForAccountingOracle;
+        _hashConsensusForValidatorsExitBusOracle = _config.hashConsensusForValidatorsExitBusOracle;
+        _gateSeal = _config.gateSeal;
+        _withdrawalCredentials = _config.withdrawalCredentials;
+        _nodeOperatorsRegistryStuckPenaltyDelay = _config.nodeOperatorsRegistryStuckPenaltyDelay;
+        _withdrawalQueueImplementation = _configImpl.withdrawalQueueImplementation;
+        _stakingRouterImplementation = _configImpl.stakingRouterImplementation;
+        _accountingOracleImplementation = _configImpl.accountingOracleImplementation;
+        _validatorsExitBusOracleImplementation = _configImpl.validatorsExitBusOracleImplementation;
+        _dummyImplementation = _configImpl.dummyImplementation;
+        _locatorImplementation = _configImpl.locatorImplementation;
+    }
+
+    function verifyInitialState() external view {
+        _verifyInitialState();
+    }
+
+    /// Need to be called before LidoOracle implementation is upgraded to LegacyOracle
+    function startUpgrade() external {
+        _startUpgrade();
+    }
+
+    function finishUpgrade() external {
+        _finishUpgrade();
+    }
+
+    /// Perform basic checks to revert the entire upgrade if something gone wrong
+    function verifyFinishedUpgrade() external view {
+        _verifyFinishedUpgrade();
+    }
+
+    function _startUpgrade() internal {
+        if (msg.sender != _voting) revert OnlyVotingCanUpgrade();
+        if (isUpgradeStarted) revert CanOnlyStartOnce();
+        isUpgradeStarted = true;
+
+        // Need to check / set locator implementation first because the other initial state checks depend on correct locator
+        if (IOssifiableProxy(address(_locator)).proxy__getImplementation() != _locatorImplementation) {
+            IOssifiableProxy(address(_locator)).proxy__upgradeTo(_locatorImplementation);
+        }
+
+        _verifyInitialState();
+
+        _prepareAccountingOracle();
+
+        _prepareValidatorsExitBus();
+
+        _migrateLidoOracleCommitteeMembers();
+
+        _prepareWithdrawalQueue();
+
+        _prepareStakingRouter();
+
+        _migrateDSMGuardians();
+    }
+
+    function _verifyInitialState() internal view {
+        if (ILidoOracle(_locator.legacyOracle()).getVersion() != 3) revert LidoOracleMustNotBeUpgradedToLegacyYet();
+
+        _verifyProxyAdmins(address(this));
+
+        // TODO: put back if enough gas
+        _verifyInitialProxyImplementations();
+
+        _verifyOZAdmins(address(this));
+
+        if (IDepositSecurityModule(_locator.depositSecurityModule()).getOwner() != address(this)) revert WrongDsmOwner();
+    }
+
+    function _verifyProxyAdmins(address admin) internal view {
```

Comment:
+ withdrawalVault?

Reply:
we have to have the upgradable contract for the withdrawal credentials till triggerable exits are implemented